datacube/index/eo3.py

```diff
@@ -273,3 +283,330 @@ def lineage_remap(name: str, uuids) -> dict[str, Any]:
+            source = RasterDatasetDataSource(band)
+            with source.open() as rdr:
+                geoboxes[name] = GeoBox(rdr.shape, rdr.transform, CRS(str(rdr.crs)))
+    else:
+        for name, m in ds.measurements.items():
+            shape, transform = _shape_and_transform(m.get("cell_size"), m.get("shape"))
+            if shape is not None:
+                assert transform is not None
+                geoboxes[name] = GeoBox(shape, transform, ds.crs)
+
+    # no grid info in measurements, see if we can get a default from browse or product
+    if geoboxes == {}:
+        full = get_in(["browse", "full"], ds.metadata_doc, default={})
+        if "cell_size" in full or "shape" in full:
+            shape, transform = _shape_and_transform(
+                full.get("cell_size"), full.get("shape"), True
+            )
+        elif ds.product.default_resolution:
+            shape, transform = _shape_and_transform(
+                {
+                    "x": ds.product.default_resolution.x,
+                    "y": ds.product.default_resolution.y,
+                }
+            )
+        elif ds.product.grid_spec:
+            # assumes we'll get a legacy GridSpec since we're dealing with legacy datasets
+            shape, transform = _shape_and_transform(
+                {
+                    "x": ds.product.grid_spec.resolution[1],
+                    "y": ds.product.grid_spec.resolution[0],
+                }
+            )
+        else:
+            raise ValueError(
+                "Unable to retrieve resolution or shape values necessary for calculating the dataset grids.\n"
+                "You may want to try again with open_datafiles=True to retrieve the information from the band files, "
+                "but be warned that it may be very slow.",
+            )
+        return {"default": {"shape": shape, "transform": transform[:6]}}, {}  # type: ignore[index]
+
+    named_gboxes, band2grid = _group_geoboxes(geoboxes)
+    grids = {
+        name: {"shape": gbox.shape.yx, "transform": gbox.transform[:6]}  # type: ignore[index]
+        for name, gbox in named_gboxes.items()
+    }
+    return grids, band2grid
+
+
+def convert_bands(
+    ds: Dataset, grid_mappings: dict[str, str]
+) -> dict[str, dict[str, str]]:
+    def _to_measurement(name, band):
+        m = {"path": band.get("path", "")}
+        if grid_mappings.get(name, "default") != "default":
+            m["grid"] = grid_mappings[name]
+        if "label" in band:
+            m["aliases"] = [band.get("label")]
+        # others?
+        return m
+
+    return {name: _to_measurement(name, band) for name, band in ds.measurements.items()}
+
+
+def convert_eo_dataset(eo_ds: Dataset, open_datafiles: bool = False) -> Dataset:
+    # Update the metadata doc of an eo-type Dataset to be eo3-compatible
+    if eo_ds.is_eo3:
+        return eo_ds
+
+    if eo_ds._gs is None or "extent" not in eo_ds.metadata_doc:
+        raise ValueError(
```

Comment:
It would be nice to raise more specific Errors here. We're not doing it at all consistently elsewhere, but... :)

Define a new Exception class and use it here, unless there's one that fits already. Something like:
```
from datacube.utils import DatacubeException
class EO3ConversionError(DatacubeException):
    """Raised when attempting and unable to convert a legacy EO dataset to EO3"""
```

It could even be fancier and have the unconvertable dataset attached to it, but,